testbeam_analysis/result_analysis.py

```diff
@@ -1299,3 +1299,273 @@ def histogram_track_angle(input_tracks_file, input_alignment_file=None, output_t
+        return np.sqrt(pitch * pitch * perc_cluster_size_1 / 100.)
+
+    def normalize_distributions(hists, fit_results):
+        '''Normalize CS distributions in order to compare them.
+        '''
+        hist_normed = [hists[0] - fit_results[0][3] + fit_results[2][0] + fit_results[2][3],  # shift CS 1 distribution (x-axis projection) such that backgrounds of CS 2 and CS 1 is the same
+                       hists[1],  # CS 2 distribution (x-axis projection) stays the same
+                       hists[2] - fit_results[1][3] + fit_results[3][0] + fit_results[3][3],  # shift CS 1 distribution (y-axis projection) such that backgrounds of CS 2 and CS 1 is the same
+                       hists[3]]  # CS 2 distribution (y-axis projection) stays the same
+        return hist_normed
+
+    # read alignment for converting the track intersections with the respective dut (global reference system) to local track intersections
+    with tb.open_file(input_alignment_file, mode="r") as in_file_h5:  # Open file with alignment data
+        if use_prealignment:
+            logging.info('Use pre-alignment data')
+            prealignment = in_file_h5.root.PreAlignment[:]
+            n_duts = prealignment.shape[0]
+        else:
+            logging.info('Use alignment data')
+            alignment = in_file_h5.root.Alignment[:]
+            n_duts = alignment.shape[0]
+
+    use_duts = use_duts if use_duts is not None else range(n_duts)  # standard setting: fit tracks for all DUTs
+
+    if plot:
+        output_pdf = PdfPages(output_pdf_file + '.pdf')
+    else:
+        output_pdf = None
+
+    n_bins = [n_bins, n_bins] if not isinstance(n_bins, Iterable) else n_bins
+
+    effective_pitches = []
+
+    with tb.open_file(input_tracks_file, 'r') as in_file_h5:
+        for actual_dut in use_duts:
+            node = in_file_h5.get_node(in_file_h5.root, 'Tracks_DUT_%d' % actual_dut)
+
+            actual_pixel_size = pixel_size[actual_dut]
+
+            initialize = True
+            for tracks_chunk, _ in analysis_utils.data_aligned_at_events(node, chunk_size=chunk_size):  # read track file in chunks
+                print tracks_chunk.dtype
+                # select only valid track intersection
+                tracks_chunk = tracks_chunk[~np.isnan(tracks_chunk[:]['offset_0'])]
+
+                # select different CS in order to histogram them and to deduce from the CS ratio the effective CS-1-pitch
+                n_hits = tracks_chunk[~np.isnan(tracks_chunk['n_hits_dut_%d' % actual_dut])]['n_hits_dut_%d' % actual_dut]
+
+                # read track intersections wiht actual dut for cluster sizes betwenn 1 and 4
+                intersection_x, intersection_y, intersection_z = tracks_chunk[:]['offset_0'], tracks_chunk[:]['offset_1'], tracks_chunk[:]['offset_2']
+                intersection_x_cs_1 = intersection_x[np.where(tracks_chunk[:]['n_hits_dut_%i' % actual_dut] == 1)]
```

Comment:
`np.where` creates new arrays and is therefore slow. We should not use it in our heavy data crunching code. Better are [boolean mask arrays](https://docs.scipy.org/doc/numpy/user/basics.indexing.html#boolean-or-mask-index-arrays) that create view of the original data. Thus the line should be: 
`intersection_x_cs_1 = intersection_x[tracks_chunk['n_hits_dut_%i' % actual_dut] == 1]`
which is also better to read.